File: lib/constructs/security/iam-construct.ts
import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kms from 'aws-cdk-lib/aws-kms';
import { SecurityConfig } from '../../config/security-config';

/**
 * IAM Construct implementing least privilege access principles
 * Creates roles and policies with minimal required permissions
 */
export class IamConstruct extends Construct {
  public readonly ec2Role: iam.Role;
  public readonly cloudTrailRole: iam.Role;
  public readonly configRole: iam.Role;
  public readonly applicationRole: iam.Role;

  constructor(
    scope: Construct,
    id: string,
    kmsKeys: {
      s3Key: kms.Key;
      secretsKey: kms.Key;
      cloudTrailKey: kms.Key;
    }
  ) {
    super(scope, id);

    // EC2 Instance Role with minimal required permissions
    this.ec2Role = new iam.Role(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-EC2-Role`,
      {
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        description: 'IAM role for EC2 instances with least privilege access',

        // Attach AWS managed policies for basic functionality
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'CloudWatchAgentServerPolicy'
          ),
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'AmazonSSMManagedInstanceCore'
          ), // For Systems Manager
        ],

        // Custom inline policy for specific application needs
        inlinePolicies: {
          [`${SecurityConfig.RESOURCE_PREFIX}-EC2-Policy`]:
            new iam.PolicyDocument({
              statements: [
                // Allow reading from specific S3 buckets only
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: ['s3:GetObject', 's3:GetObjectVersion'],
                  resources: [
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-*/*`,
                  ],
                }),

                // Allow decryption of application-specific KMS keys
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    'kms:Decrypt',
                    'kms:GenerateDataKey',
                    'kms:DescribeKey',
                  ],
                  resources: [kmsKeys.s3Key.keyArn, kmsKeys.secretsKey.keyArn],
                }),

                // Allow reading specific secrets
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    'secretsmanager:GetSecretValue',
                    'secretsmanager:DescribeSecret',
                  ],
                  resources: [
                    `arn:aws:secretsmanager:${SecurityConfig.PRIMARY_REGION}:*:secret:${SecurityConfig.RESOURCE_PREFIX}/*`,
                  ],
                }),
              ],
            }),
        },
      }
    );

    // Application-specific role for Lambda functions or containers
    this.applicationRole = new iam.Role(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-App-Role`,
      {
        assumedBy: new iam.CompositePrincipal(
          new iam.ServicePrincipal('lambda.amazonaws.com'),
          new iam.ServicePrincipal('ecs-tasks.amazonaws.com')
        ),
        description:
          'IAM role for application services with restricted permissions',

        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AWSLambdaVPCAccessExecutionRole'
          ),
        ],

        inlinePolicies: {
          [`${SecurityConfig.RESOURCE_PREFIX}-App-Policy`]:
            new iam.PolicyDocument({
              statements: [
                // Restricted S3 access
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: ['s3:GetObject', 's3:PutObject', 's3:DeleteObject'],
                  resources: [
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-app-data/*`,
                  ],
                }),

                // CloudWatch Logs access
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    'logs:CreateLogGroup',
                    'logs:CreateLogStream',
                    'logs:PutLogEvents',
                  ],
                  resources: [
                    `arn:aws:logs:${SecurityConfig.PRIMARY_REGION}:*:log-group:/aws/lambda/${SecurityConfig.RESOURCE_PREFIX}*`,
                  ],
                }),
              ],
            }),
        },
      }
    );

    // CloudTrail Service Role
    this.cloudTrailRole = new iam.Role(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-CloudTrail-Role`,
      {
        assumedBy: new iam.ServicePrincipal('cloudtrail.amazonaws.com'),
        description: 'IAM role for CloudTrail service',

        inlinePolicies: {
          [`${SecurityConfig.RESOURCE_PREFIX}-CloudTrail-Policy`]:
            new iam.PolicyDocument({
              statements: [
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    's3:PutObject',
                    's3:GetBucketAcl',
                    's3:PutBucketAcl',
                  ],
                  resources: [
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-cloudtrail-logs`,
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-cloudtrail-logs/*`,
                  ],
                }),
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: ['kms:GenerateDataKey*', 'kms:DescribeKey'],
                  resources: [kmsKeys.cloudTrailKey.keyArn],
                }),
              ],
            }),
        },
      }
    );

    // AWS Config Service Role
    this.configRole = new iam.Role(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-Config-Role`,
      {
        assumedBy: new iam.ServicePrincipal('config.amazonaws.com'),
        description: 'IAM role for AWS Config service',

        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AWS_ConfigRole'
          ),
        ],

        inlinePolicies: {
          [`${SecurityConfig.RESOURCE_PREFIX}-Config-Policy`]:
            new iam.PolicyDocument({
              statements: [
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    's3:GetBucketAcl',
                    's3:ListBucket',
                    's3:GetBucketLocation',
                    's3:PutObject',
                    's3:GetBucketVersioning',
                  ],
                  resources: [
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-config-logs`,
                    `arn:aws:s3:::${SecurityConfig.RESOURCE_PREFIX.toLowerCase()}-config-logs/*`,
                  ],
                }),
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                    'kms:GenerateDataKey*',
                    'kms:DescribeKey',
                    'kms:Decrypt',
                  ],
                  resources: [kmsKeys.cloudTrailKey.keyArn],
                }),
              ],
            }),
        },
      }
    );

    // Create instance profile for EC2
    new iam.InstanceProfile(
      this,
      `${SecurityConfig.RESOURCE_PREFIX}-EC2-InstanceProfile`,
      {
        role: this.ec2Role,
        instanceProfileName: `${SecurityConfig.RESOURCE_PREFIX}-EC2-InstanceProfile`,
      }
    );

    // Apply tags to all IAM resources
    const roles = [
      this.ec2Role,
      this.applicationRole,
      this.cloudTrailRole,
      this.configRole,
    ];
    roles.forEach(role => {
      Object.entries(SecurityConfig.STANDARD_TAGS).forEach(([key, value]) => {
        role.node.addMetadata(key, value);
      });
    });
  }
}

File: archive/cdk-java/Pr2382/lib/src/main/java/app/constructs/IamConstruct.java
package app.constructs;

import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.amazon.awscdk.services.iam.PolicyDocument;
import software.amazon.awscdk.services.iam.PolicyStatement;
import software.amazon.awscdk.services.iam.Effect;
import software.constructs.Construct;
// Removed imports for account-scoped custom resources to avoid runtime AWS account calls during synth
import software.amazon.awscdk.services.iam.ManagedPolicy;
import app.config.EnvironmentConfig;
import java.util.List;
import java.util.Map;
import software.amazon.awscdk.services.iam.Group;
import software.amazon.awscdk.services.iam.ArnPrincipal;
import software.amazon.awscdk.services.iam.User;

/**
 * IAM construct that creates roles and policies following the principle of least privilege.
 * This construct also enforces strong password policies and MFA requirements
 * for enhanced security in financial services environments.
 */
public class IamConstruct extends Construct {
    
    private final Role s3ReadOnlyRole;
    private final Role cloudTrailRole;
    
    public IamConstruct(final Construct scope, final String id) {
        super(scope, id);
        // Account password policy (PutAccountPasswordPolicy) previously used an
        // SDK-backed custom resource which performs account-level API calls during
        // synth. To avoid synth-time account calls and pipeline failures, that
        // behavior was removed from this construct. Account-level governance
        // should be managed separately (CI/CD or org policy).
        
        // Create IAM roles with least privilege
        this.s3ReadOnlyRole = createS3ReadOnlyRole();
        this.cloudTrailRole = createCloudTrailRole();
    }

    /**
     * Creates a group for human operators and attaches the MFA enforcement policy.
     */
    // Operators group and MFA enforcement removed.
    
    /**
     * Creates a strict password policy for IAM users.
     * Enforces 90-day rotation, complexity requirements, and prevents password reuse.
     */
    
    /**
     * Creates an IAM role for S3 read-only access with least privilege principle.
     */
    private Role createS3ReadOnlyRole() {
        return Role.Builder.create(this, EnvironmentConfig.getResourceName("iam", "s3-readonly-role"))
                .assumedBy(new ServicePrincipal("ec2.amazonaws.com"))
                .description("Role for EC2 instances requiring read-only S3 access")
                .inlinePolicies(Map.of(
                    "S3ReadOnlyPolicy", PolicyDocument.Builder.create()
                        .statements(List.of(
                            PolicyStatement.Builder.create()
                                .effect(Effect.ALLOW)
                                .actions(List.of(
                                    "s3:GetObject",
                                    "s3:GetObjectVersion",
                                    "s3:ListBucket"
                                ))
                                .resources(List.of(
                                    String.format("arn:aws:s3:::%s-*", EnvironmentConfig.ENVIRONMENT),
                                    String.format("arn:aws:s3:::%s-*/*", EnvironmentConfig.ENVIRONMENT)
                                ))
                                .build()
                        ))
                        .build()
                ))
                .build();
    }
    
    /**
     * Creates an IAM role for CloudTrail service with minimal required permissions.
     */
    private Role createCloudTrailRole() {
        return Role.Builder.create(this, EnvironmentConfig.getResourceName("iam", "cloudtrail-role"))
                .assumedBy(new ServicePrincipal("cloudtrail.amazonaws.com"))
                .description("Role for CloudTrail logging service")
                .inlinePolicies(Map.of(
                    "CloudTrailLogsPolicy", PolicyDocument.Builder.create()
                        .statements(List.of(
                            PolicyStatement.Builder.create()
                                .effect(Effect.ALLOW)
                                .actions(List.of(
                                    "logs:CreateLogGroup",
                                    "logs:CreateLogStream",
                                    "logs:PutLogEvents",
                                    "logs:DescribeLogGroups",
                                    "logs:DescribeLogStreams"
                                ))
                                // Limit to CloudWatch Logs in this account; CloudTrail writes to log groups with the pattern /aws/cloudtrail/*
                                .resources(List.of(String.format("arn:aws:logs:*:%s:log-group:/aws/cloudtrail/*", software.amazon.awscdk.Stack.of(this).getAccount())))
                                .build()
                        ))
                        .build()
                ))
                .build();
    }

    /**
     * Create an account password policy using the L1 construct so the account
     * enforces 90-day rotation and complexity rules.
     */
    private void createPasswordPolicy() {
    // Account-level password policy application removed.
    // If desired, apply account password policies via an organizational
    // automation or in a deployment step that has explicit credentials and
    // the proper permissions. Keeping this method as a no-op avoids
    // synth-time account API calls that break CI.
    }

    /**
     * Creates a managed policy that can be attached to user groups to enforce MFA
     * and limit actions for human users. This is intentionally conservative.
     */
    private ManagedPolicy createMfaEnforcementPolicy() {
    // MFA enforcement removed per request.
    return null;
    }
    
    // Getters
    public Role getS3ReadOnlyRole() { return s3ReadOnlyRole; }
    public Role getCloudTrailRole() { return cloudTrailRole; }
    // Operators group removed.
    // Password policy getter intentionally omitted; refer to account-level governance
}